application/apps/text_grep/src/wc_1_ahocorasick.rs

```diff
@@ -0,0 +1,165 @@
+    let patterns_arc: Vec<_> = patterns.iter().map(|&p| Arc::from(p)).collect();
+    let file_paths_arc: Vec<_> = file_paths.iter().map(|&fp| Arc::from(fp)).collect();
+
+    // spawn a thread for each file to read it concurrently
+    let thread_handles: Vec<_> = file_paths_arc
+      .iter()
+      .map(|file_path| {
+        let patterns = patterns_arc.clone();
+        let sender = sender.clone();
+        let error_sender_clone = error_sender.clone();
+        let cancel_token = cancel_token_clone.clone();
+        let file_path = Arc::clone(file_path);
+        thread::spawn(move || {
+          if let Err(err) = process_file(&file_path, &patterns, chunk_size, &cancel_token, &sender) {
+            if error_sender_clone.send(err.to_string()).is_err() {
+              eprintln!("Error sending error message through channel");
+            }
+          }
+        })
+      })
+      .collect();
+
+    eprintln!("Reached till before join of handles");
+    // Wait for all threads to finish
+    for handle in thread_handles {
+      handle.join().unwrap();
+    }
+
+    eprintln!("Reached till after join of handles");
+
+    while let Ok(err_msg) = error_receiver.try_recv() {
+      eprintln!("Error processing file: {:?}", err_msg);
+      // Push a SearchResult with the error message
+      results.push(SearchResult {
+        file_path: "".to_string(),
+        pattern_counts: HashMap::new(),
+        error_message: Some(err_msg.to_string()),
+      });
+    }
+
+    // Loop to receive search results
+    while let Ok(search_result) = receiver.try_recv() {
+      results.push(search_result?);
+    }
+
+    Ok(results)
+  }
+}
+
+fn process_file(
+  file_path: &Arc<str>,
+  patterns: &[Arc<str>],
+  chunk_size: usize,
+  cancel_token: &CancellationToken,
+  sender: &mpsc::Sender<Result<SearchResult, String>>,
+) -> Result<(), String> {
+  let file_path = PathBuf::from(&**file_path);
+  if !is_text_file(&file_path) {
+    let error_msg = format!("File '{}' is not a text file", file_path.display());
+    if sender.send(Err(error_msg.clone())).is_err() {
+      eprintln!("Error sending search result through channel");
+    }
+    return Ok(());
+  }
+
+  let mut pattern_counts = HashMap::new();
+  let start_time = Instant::now();
+  eprintln!("Trying to process file {:?}", file_path);
+  // Preparing searchers for all patterns
+  let searchers: Vec<_> = patterns
+    .iter()
+    .map(|pattern| {
+      aho_corasick::AhoCorasickBuilder::new()
+        .match_kind(aho_corasick::MatchKind::Standard)
+        .ascii_case_insensitive(true)
+        .build(vec![&**pattern])
+    })
+    .collect::<Result<_, _>>()
+    .map_err(|e| e.to_string())?;
+
+  eprintln!("Trying to Open file {:?}", file_path);
+  let file = File::open(&file_path).map_err(|e| e.to_string())?;
+  let mut reader = BufReader::new(file);
+  let mut buffer = vec![0; chunk_size];
+
+  loop {
+    if cancel_token.is_cancelled() {
+      return Err("Operation cancelled".to_string());
```

Comment:
I don't think in case of cancellation, we should return error... Probably `Result<Option<...>, E>` with `Ok(None)` in case of cancellation would be better